pkg-r/R/contents_shinychat.R

```diff
@@ -1,14 +1,296 @@
-  # TODO(garrick): placeholder, will be replaces with S7 generic
-  check_ellmer_chat(client)
+opt_shinychat_tool_display <- function() {
+  choices <- c("none", "basic", "rich")
 
-  client_msgs <- map(client$get_turns(), function(turn) {
-    content <- ellmer::contents_markdown(turn)
+  opt <- getOption("shinychat.tool_display", default = NULL)
+  if (!is.null(opt)) {
+    opt <- arg_match(opt, choices, error_arg = "shinychat.tool_display")
+    return(opt)
+  }
+
+  env <- Sys.getenv("SHINYCHAT_TOOL_DISPLAY", unset = "rich")
+  arg_match(env, choices, error_arg = "SHINYCHAT_TOOL_DISPLAY")
+}
+
+#' Format ellmer content for shinychat
+#'
+#' @param content An [`ellmer::Content`] object.
+#'
+#' @return Returns text, HTML, or web component tags formatted for use in
+#'   `chat_ui()`.
+#'
+#' @export
+contents_shinychat <- S7::new_generic(
+  "contents_shinychat",
+  "content",
+  function(content) {
+    S7::S7_dispatch()
+  }
+)
+
+S7::method(contents_shinychat, ellmer::Content) <- function(content) {
+  # Fall back to html or markdown
+  html <- ellmer::contents_html(content)
+  if (!is.null(html)) {
+    shiny::HTML(html)
+  } else {
+    ellmer::contents_markdown(content)
+  }
+}
+
+S7::method(contents_shinychat, ellmer::ContentText) <- function(content) {
+  content@text
+}
+
+S7::method(contents_shinychat, ellmer::ContentToolRequest) <- function(
+  content
+) {
+  if (opt_shinychat_tool_display() == "none") {
+    return(NULL)
+  }
+
+  props <- list(
+    "request-id" = content@id,
+    name = content@name,
+    arguments = jsonlite::toJSON(content@arguments, auto_unbox = TRUE),
+    intent = content@arguments$.intent
+  )
+
+  tool <- content@tool
+  if (!is.null(tool)) {
+    props$title <- tool@annotations$title
+  }
+
+  tagList(
+    htmltools::tag("shiny-tool-request", props),
+    chat_deps()
+  )
+}
+
+S7::method(contents_shinychat, ellmer::ContentToolResult) <- function(content) {
+  if (opt_shinychat_tool_display() == "none") {
+    return(NULL)
+  }
+
+  if (is.null(content@request)) {
+    cli::cli_abort(
+      "`ContentToolResult` objects must have an associated `@request` property."
+    )
+  }
+
+  content@extra$display <- ensure_content_display(content)
+
+  # Prepare base props
+  props <- list(
+    request_id = content@request@id,
+    request_call = "",
+    name = content@request@name,
+    status = if (tool_errored(content)) "error" else "success",
+    show_request = if (!isFALSE(content@extra$display$show_request)) NA,
+    intent = content@request@arguments$.intent,
+    expanded = if (isTRUE(content@extra$display$open)) NA
+  )
+
+  icon_deps <- NULL
+  tool <- content@request@tool
+
+  # Let tool results override global tool properties
+  if (!is.null(content@extra$display$title)) {
+    props$title <- content@extra$display$title
+  }
+  if (!is.null(content@extra$display$icon)) {
+    props$icon <- content@extra$display$icon
+    icon_deps <- htmltools::findDependencies(props$icon)
+  }
```

Comment:
Similar note to https://github.com/posit-dev/shinychat/pull/52/files#r2252686730